src/main/java/net/neoforged/neoforge/event/AddBlockToBlockEntityTypeEvent.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.event;
+
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.SignBlock;
+import net.minecraft.world.level.block.StandingSignBlock;
+import net.minecraft.world.level.block.WallSignBlock;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.neoforged.bus.api.Event;
+import net.neoforged.fml.ModLoader;
+import net.neoforged.fml.event.IModBusEvent;
+import org.jetbrains.annotations.Nullable;
+
+public class AddBlockToBlockEntityTypeEvent extends Event implements IModBusEvent {
+    private static final MethodHandle VALID_BLOCKS_SETTER_METHOD_HANDLE;
```

Comment:
This makes code harder to reason about. Is the alternative that much worse?

Reply:
The alternative is exposing the validBlocks field with an AT or public getter. Which is exactly what we are trying to avoid. Alternative is reflection I guess but method handles should be easily optimized as what Luke said above